code/modules/xenoarchaeology/xenoartifact_structure.dm

```diff
@@ -0,0 +1,277 @@
+	charge_req = X.charge_req*1.5
+	special_desc = X.special_desc
+	touch_desc = X.touch_desc
+	alpha = X.alpha
+	traits = X.traits
+	little_man_inside_me = X
+	little_man_inside_me.forceMove(src)
+
+	for(var/datum/xenoartifact_trait/T in traits)
+		if(!istype(T, /datum/xenoartifact_trait/minor/dense))
+			T.on_init(src)
+
+	var/holdthisplease = rand(1, 3)
+	icon_state = "SB[holdthisplease]"//Base
+	generate_icon(icon, "SBL[holdthisplease]", material)
+	if(prob(70) || icon_slots[1])//Top
+		if(!(icon_slots[1])) //Some traits can set this too, it will be set to a code that looks like 9XX
+			icon_slots[1] = rand(1, 3)
+		generate_icon(icon, "ST[icon_slots[1]]")
+		generate_icon(icon, "STL[icon_slots[1]]", material)
+		
+		if(prob(70) || icon_slots[2])//Bottom
+			if(!(icon_slots[2]))
+				icon_slots[2] = rand(1, 3)
+			generate_icon(icon, "SBTM[icon_slots[2]]")
+			generate_icon(icon, "SBTML[icon_slots[2]]", material)
+
+	if(prob(50)  || icon_slots[3])//Left
+		if(!(icon_slots[3]))
+			icon_slots[3] = rand(1, 2)
+		generate_icon(icon, "SL[icon_slots[3]]")
+		generate_icon(icon, "SLL[icon_slots[3]]", material)
+
+	if(prob(50) || icon_slots[4])//Right
+		if(!(icon_slots[4]))
+			icon_slots[4] = rand(1, 2)
+		generate_icon(icon, "SR[icon_slots[4]]")
+		generate_icon(icon, "SRL[icon_slots[4]]", material)
+
+/obj/structure/xenoartifact/examine(mob/living/carbon/user)
+	. = ..()
+	if(istype(user.glasses, /obj/item/clothing/glasses/science))
+		to_chat(user, "<span class='notice'>[special_desc]</span>")
+
+/obj/structure/xenoartifact/attack_hand(mob/user)
+	. = ..()
+	if(process_type == LIT) //Snuff out candle
+		to_chat(user, "<span class='notice'>You snuff out [name]</span>")
+		process_type = null
+		return
+	if(user.a_intent == INTENT_GRAB)
+		touch_desc?.on_touch(src, user)
+		return
+	SEND_SIGNAL(src, XENOA_INTERACT, null, user, user)
+
+/obj/structure/xenoartifact/throw_impact(atom/target, mob/user)
+	. = ..()
+	SEND_SIGNAL(src, XENOA_THROW_IMPACT, null, user, target)
+
+/obj/structure/xenoartifact/attackby(obj/item/I, mob/living/user, params)
+	for(var/datum/xenoartifact_trait/T in traits)
+		T.on_item(src, user, I)
+	if(!(manage_cooldown(TRUE))||user?.a_intent == INTENT_GRAB||istype(I, /obj/item/xenoartifact_label)||istype(I, /obj/item/xenoartifact_labeler))
+		return
+	SEND_SIGNAL(src, XENOA_ATTACKBY, I, user, user)
+	..()
+
+/*
+	check_charge() is essentially what runs all the minor, major, and malf trait activations. 
+	This process also culls any irrelivent targets in reference to max_range and calculates the true charge.
+	True charge is simply, almost, the average of the charge and charge_req. This allows for a unique varience of 
+	output from artifacts, generally producing some funny results too.
+	
+*/
+/obj/structure/xenoartifact/proc/check_charge(mob/user, charge_mod)
+	if(prob(malfunction_chance)) //See if we pick up an malfunction
+		var/datum/xenoartifact_trait/T = pick(subtypesof(/datum/xenoartifact_trait/malfunction))
```

Comment:
subtypeof call on check_charge?

Reply:
Does that impact performance badly? It only enters if it rolls successfully for a malfunction.